### PR TITLE
Exercise 1

### PR DESCRIPTION
**Pull request by danmousley367**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/templates/index.html [hunk 1]
```diff
@@ -6,13 +6,30 @@
...
+    <table class="table">
+      <tr>
+          <th>Item</th>
+          <th>Status</th>
+          <th></th>
+          <th></th>
+      </tr>
+      {% for item in items %}
+      <tr>
+          <td>{% if item.status == "Finished" %}<del>{{item.title}}</del>{% else %}{{item.title}}{% endif %}</td>
```
Comment (JackMead) [inline todo_app/templates/index.html:27]: Minor, but some of these lines are getting quite long - is it worth spreading them out a little?
```html
<td>
            {% if item.status == "Finished" %}
              <del>{{item.title}}</del>
            {% else %}
              {{item.title}}
            {% endif %}
          </td>
```

Comment (JackMead) [inline todo_app/templates/index.html:27]: Minor, but some of these lines are getting quite long - is it worth spreading them out a little?
```html
          <td>
            {% if item.status == "Finished" %}
              <del>{{item.title}}</del>
            {% else %}
              {{item.title}}
            {% endif %}
          </td>
```

File: todo_app/app.py [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
+                  deleted_item = get_item(deleted_item_id)
+
+                  delete_item(deleted_item)
+
+            return redirect(url_for('index'))
+
+    items = get_items()
+    sorted_items = sorted(items, key=lambda item: item['status'], reverse=True) 
+
+    return render_template('index.html', items=sorted_items)
```
Comment (JackMead) [inline todo_app/app.py:38]: This function is getting quite long - would it be worth splitting this up a bit?
E.g. I'd be tempted to split the two options (change status, delete) into separate forms with separate routes, and splitting the POST from the GET, e.g. to start we might just break POST & GET:
```python
@app.route('/')
def index():
    items = get_items()
    sorted_items = sorted(items, key=lambda item: item['status'], reverse=True) 

    return render_template('index.html', items=sorted_items)

@app.route('/', methods=['POST'])
def change_item():
      if 'added_item' in request.form.keys():
            item_title = request.form['added_item']
            add_item(item_title)

      if 'updated_item' in request.form.keys():
            updated_item_id = request.form['updated_item']
            updated_item = get_item(updated_item_id)

            if updated_item['status'] == 'Not Started':
                  updated_item['status'] = 'Finished'
            elif updated_item['status'] == 'Finished':
                  updated_item['status'] = 'Not Started'
            save_item(updated_item)

      if 'deleted_item' in request.form.keys():
            deleted_item_id = request.form['deleted_item']
            deleted_item = get_item(deleted_item_id)

            delete_item(deleted_item)

      return redirect(url_for('index'))
```
And then we could start to have separate POSTs:
```python
@app.route('/add', methods=["POST"])
def add_new():
   pass

@app.route('/update', methods=["POST"])
def update():
    pass
```
etc.

**Review by JackMead — APPROVED** [review]
Nice work, you've got an effective solution that I had no problems setting up & running to view/manage my to-dos, and well done on the various stretch goals too! 😄 I've left a couple of optional thoughts below, but very happy for you to merge this when you're ready

File: todo_app/templates/index.html (hunk 1 above)
Comment (danmousley367) [reply todo_app/templates/index.html:27]: Good shout, will do!

File: todo_app/app.py (hunk 2 above)
Comment (danmousley367) [reply todo_app/app.py]: Thanks! I've split up GET and POST. I quite like having everything on the one page, unless I'm misunderstanding how the different routes work?

Comment (JackMead) [reply todo_app/app.py]: Definitely fine to have everything on one page (definitely my preferred lazy approach for these exercises!) but we can split up the POST requests as much as we want and still achieve that, e.g. by having two forms (which might be nice anyway) and specifying separate actions:
```html
        <form action="/add" method="post">
            <button type=submit class="btn btn-secondary" name="updated_item" value={{item.id}}>
              Mark {% if item.status == "Finished" %}incomplete{% else %}complete{% endif %}
            </button>
        </form>
```
and
```html
        <form action="/delete" method="post">
            <button type=submit class="btn btn-danger" name="deleted_item" value={{item.id}}>
              x
            </button>
        </form>
```

Comment (danmousley367) [reply todo_app/app.py]: Ahh that makes sense, I missed the part about specifying an action on the form - have done this in exercise 2. Thanks Jack!

**Pull request merged**